### ISSUE TITLE
plan

### ISSUE THREAD

**Issue by ashstarry** [ashstarry/cloud_computing-labs#4]
Design Document: Deprecating createMesh in AWS App Mesh
Objective
To outline the process of deprecating the createMesh function in AWS App Mesh, allowing only specific customers to access this feature.

Background
AWS App Mesh is intended to be deprecated due to [provide reasons]. To ensure a smooth transition for our customers and maintain business continuity for key accounts, we propose to block the createMesh function with exceptions.

Options
AWS Organizations-based Allowance: Leverage AWS Organizations to create permission boundaries.
Tag-based Allowance: Implement a specific tag as a "passcode" to control the access to the function.
AWS Organizations-based Allowance
Design
Create a new Organization Unit (OU): For customers allowed to use createMesh.

Service Control Policies (SCP): Implement SCPs that exclusively grant the ability to call createMesh and attach it to the newly created OU.

Onboarding: Move the accounts of the allowed customers into this new OU.

Pros
Centralized management of permissions.
Ability to control other AWS services/functions with the same approach in the future.
Cons
Requires customers to be part of the organization.
Onboarding/Offboarding overhead.
Tag-based Allowance
Design
Tag Specification: Define a specific tag (e.g., "AllowCreateMesh": "true") that acts as a passcode.

IAM Policy: Update the IAM policy associated with App Mesh to include a condition that checks for the tag's existence when createMesh is invoked.

Customer Onboarding: Communicate to selected customers about the need to apply this tag to relevant resources.

Pros
More flexible as it doesn’t require AWS Organization membership.
Simplified onboarding process.
Cons
Security concern if the tag information gets leaked.
Needs continuous monitoring to ensure only the authorized entities use the tag.
Recommendation
[Based on business requirements, user size, customer feedback, and technical feasibility, provide a recommendation.]

Implementation Steps
Communication Strategy: Inform all App Mesh customers about the upcoming deprecation.

Implement Selected Allowance Method: Follow the design steps for the chosen method.

Testing: Validate the implemented restrictions and exceptions across multiple scenarios and customer accounts.

Rollout: Introduce the restriction to all AWS App Mesh users.

Feedback Loop: Create channels for customer feedback and concerns regarding the deprecation.

Monitoring: Establish monitoring and alerts to identify any unintended issues or unauthorized createMesh calls.

Timeline
Week 1-2: Communication and feedback collection
Week 3-4: Design and Implementation
Week 5: Testing
Week 6: Rollout
Week 7 onwards: Continuous Monitoring and Feedback Loop

**Comment by ashstarry** [comment]
Customer Notification:

Advanced Warning: Give customers ample time to transition away from the deprecated service.
Multiple Channels: Use email, AWS Management Console notifications, AWS official blog posts, etc., to communicate the upcoming changes.
Migration Strategy:

Alternative Solutions: Provide guidance on alternative AWS services or strategies that customers can migrate to.
Migration Tools: If possible, offer tools or scripts that can aid the migration process.
Detailed Documentation: Provide comprehensive migration guides.
Customer Support:

Dedicated Support: Offer enhanced support options, including extended hours or dedicated channels, to help customers navigate the change.
Training: Organize webinars or workshops to educate customers about alternative solutions and migration strategies.
Feedback Channels:

Surveys: Conduct surveys to understand why some customers might be more affected than others, and tailor support accordingly.
Open Forums: Create forums or dedicated AWS re:Invent sessions to discuss the change, challenges, and potential solutions.
Financial Considerations:

Cost Implications: Understand any cost implications for customers transitioning to alternative solutions.
Credits or Discounts: Consider offering credits or discounts to severely impacted customers to aid their transition.
Monitoring & Metrics:

Deprecation Metrics: Monitor the usage of createMesh before and after implementing the block to measure the effectiveness of your communication strategy.
Migration Success Metrics: Track how many customers have successfully migrated away from the deprecated service or feature.
Rollback Strategy:

Even with extensive testing, there might be unforeseen challenges. Having a rollback strategy ensures that you can revert changes if things don't go as planned.
Legal & Compliance:

Service Level Agreements (SLA) Review: Ensure that the deprecation doesn't violate any existing SLAs or contractual agreements.
Data Compliance: If customers stored data within the service, ensure they have time and means to retrieve or migrate their data in compliance with data protection regulations.
Stakeholder Communication:

Ensure that all internal stakeholders, especially customer-facing teams like Sales and Customer Support, are well-informed and prepared to handle queries.
Post-deprecation Review:

**Comment by ashstarry** [comment]
Certainly! Utilizing a tag as a "passcode" to control the creation of resources, such as the `createMesh` API in this case, has its advantages and disadvantages. Here's a breakdown:

### Pros:

1. **Flexibility**:
   - Tagging provides a way to conditionally allow or deny operations without reconfiguring IAM permissions or service settings directly.
   
2. **Simplified Onboarding**:
   - It can be easier for users to add a tag than to manage or understand more complex permission settings, especially if the number of users/groups that need access is dynamic.
  
3. **Granular Control**:
   - Tags can be combined with other conditions in IAM policies, allowing for a multi-faceted access control strategy.

4. **Easy Identification**:
   - Later, you can quickly identify which meshes were created with the "passcode" simply by querying for the tag, facilitating audits or migrations.

5. **Temporary Access**:
   - If the need arises for temporary access, it's straightforward to add or remove a tag, compared to potentially reconfiguring IAM roles or organizational units.

### Cons:

1. **Security Risks**:
   - If the "passcode" tag becomes widely known, it could be misused, thereby bypassing the intended restrictions.
   - There's a risk of users inadvertently sharing the tag, especially if they're unaware of its significance.

2. **Maintenance Overhead**:
   - As your system grows, managing which resources have the "pass" tag and which do not could become challenging.
   - There may also be a need to periodically change or rotate the tag to maintain security, leading to additional overhead.

3. **Potential Confusion**:
   - Some users might forget to add the tag or be unaware of its necessity, leading to denied operations and potential frustration.
   - Error messaging needs to be clear to reduce confusion, but that might indirectly expose the mechanism to a wider audience.

4. **Audit Complexity**:
   - Auditing based on tags might be more complicated than evaluating straightforward IAM policies or organizational boundaries.
  
5. **Performance Implications**:
   - Depending on the implementation, checking tags could introduce a slight delay in the API call or the resource creation process.
  
6. **IAM Policy Complexity**:
   - IAM policies that check for tags can become complicated, especially if combined with other conditions.
   
7. **Less Traditional**:
   - Using tags as a passcode is less conventional than typical IAM or organizational controls, so it may be unfamiliar or confusing to some AWS administrators.

In summary, while tagging as a passcode mechanism can provide a quick and flexible method for conditional access, it also introduces potential security risks and administrative complexities. The suitability of this method would greatly depend on the specific requirements, scale, and security posture of your organization.

**Comment by ashstarry** [comment]
AppMesh will soon cease to accept new customers. However, until our existing customers have fully migrated to ECS Service Connect or third-party platforms, we will continue to provide full support to these clients. This document will discuss the implementation strategies for the aforementioned requirements, including how to ensure that existing customers (or in emergency situations) can still use AppMesh with new accounts. This document will not cover Alternatives/Migration/Customer Communication.

**Comment by ashstarry** [comment]
API: Since only the createMesh API acts as the actual creation gateway for our service, and all CRUD APIs (excluding listMeshes) operate based on existing meshes, we will only block the createMesh API.

Mesh Sharing: We will not obstruct the current customers' permissions to share meshes; this aspect will remain unchanged.

IAM Policy: We will not modify any AppMesh Permissions.

**Comment by ashstarry** [comment]
Need for a List of Current App Mesh Accounts

Given the retention period of our serviceLog is one year, theoretically, we will obtain user activity from up to the last year. We can use the following script to retrieve a potential list.

Insight Query:

(Note: The exact query will depend on whether we intend to grant permissions exclusively to the owner account or also to consumer accounts that share the mesh.)

**Comment by ashstarry** [comment]
Ensuring Uninterrupted Service for Current Customers

As we will adopt a soft migration approach, we do not wish to hinder the use of any functionality by current customers. This includes creating/operating Mesh with new accounts. While blocking any commercial account access, we need to provide a restricted but secure entry point. Three solutions are proposed here. It's essential to note that each solution will require additional actions on the customer's part.

[Details of the three solutions can be provided next.]

By framing it this way, stakeholders will understand the importance of ensuring continued service for current customers while transitioning to the new phase of the service.

**Comment by ashstarry** [comment]
Integrating with AWS Organization for Customer Validation

After customers are added to or exist within their respective AWS organizations, we aim to authorize based on the organization's membership. Based on our discussions with the AWS Organization team, making direct determinations about whether a user is using the App Mesh Service or allowing/denying createMesh permissions based on AWS Organization membership is complex and lacks direct precedents.

App Mesh Integration with AWS Organization:

Subsequent to the integration of App Mesh with AWS Organization, the AWS Organization team has proposed a Notification-based solution. The specific implementation steps are as follows:

[Please provide the specific steps or continue with the description of the solution here.]

This approach ensures that while we integrate with AWS Organization, we're relying on notifications to determine permissions, rather than direct membership checks, easing the implementation complexity.

**Comment by ashstarry** [comment]
We will provide potential customers with an identification code (different customers (entries) can have different identification codes for easier management). Only when adding this key-value identification code during the createMesh in tag settings can a mesh be created, and the account will be added to the allowlist.

**Comment by ashstarry** [comment]
Sure, here's how you can create a mesh in AWS App Mesh with a tag using various methods:

### 1. AWS CLI:

Using the AWS CLI, you can use the `create-mesh` command:
```bash
aws appmesh create-mesh --mesh-name my-mesh --tags Key=Environment,Value=Dev
```

### 2. AWS SDK (using Python and Boto3 as an example):

```python
import boto3

appmesh = boto3.client('appmesh')

response = appmesh.create_mesh(
    meshName='my-mesh',
    tags=[
        {
            'key': 'Environment',
            'value': 'Dev'
        },
    ]
)
print(response)
```

### 3. AWS CDK (using TypeScript as an example):

```typescript
import * as appmesh from '@aws-cdk/aws-appmesh';
import * as cdk from '@aws-cdk/core';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'AppMeshStack');

new appmesh.Mesh(stack, 'MyMesh', {
    meshName: 'my-mesh',
    tags: {
        Environment: 'Dev'
    }
});
```

### 4. AWS CloudFormation:

Here's how you'd define a mesh in a CloudFormation YAML template:

```yaml
Resources:
  MyMesh:
    Type: "AWS::AppMesh::Mesh"
    Properties: 
      MeshName: "my-mesh"
      Tags: 
        - Key: "Environment"
          Value: "Dev"
```

### 5. Terraform:

Using the Terraform AWS provider, you can define an App Mesh mesh as follows:

```hcl
resource "aws_appmesh_mesh" "example" {
  name = "my-mesh"
  tags = {
    Environment = "Dev"
  }
}
```

With each of these methods, adjust the `Environment` and `Dev` values to whatever key-value pair you'd like for tagging. Always remember to have proper permissions to create App Mesh resources when using these methods.

**Comment by ashstarry** [comment]
If necessary, we can add some metrics to monitor the usage of createMesh after it is closed to the public, such as:

Allowance Grant Count: Monitor the number of accounts that have been granted the special tag-based allowance over time. This can help in assessing how frequently the feature is used and its trends (whether it's increasing or decreasing).
Accounts Added to Allowlist: A metric indicating how many accounts have been added to the allowlist after successfully creating a mesh. This can provide insights into the growth and scale of the system.

**Comment by ashstarry** [comment]
Certainly! Here's a detailed approach for granting access only to accounts on an allowlist, using AWS AppConfig for storing the allowlist and Lambda for updating the list:

### 1. Setting up AWS AppConfig:

1.1. **Create an AWS AppConfig Application**: This represents the grouping of related configuration profiles.

1.2. **Create a Configuration Profile**: This is where you'll store the list of allowed accounts. For simplicity, let's assume it's a JSON file with the following format:

```json
{
    "allowedAccounts": ["account1", "account2", ...]
}
```

### 2. Using AWS Lambda to update the allowlist:

2.1. **Lambda Function**:
- Create a new Lambda function, let's name it `UpdateAllowlistFunction`.
- Grant this function permissions to update AppConfig configuration.
  
2.2. **Lambda Handler**:
Here's a Python example for a Lambda function to add a new account to the allowlist:

```python
import boto3
import json

appconfig = boto3.client('appconfig')

def lambda_handler(event, context):
    # Assuming the event contains the account to be added
    account_to_add = event['account']

    # Fetch the current configuration from AppConfig
    response = appconfig.get_configuration(
        Application="YourAppConfigApplicationName",
        Environment="YourEnvironment",
        Configuration="YourConfigurationProfile",
        ClientId="SomeClientIdentifier" 
    )
    config_data = json.loads(response['Content'].read())



```python
    # Add the new account to the allowedAccounts list if not already present
    if account_to_add not in config_data['allowedAccounts']:
        config_data['allowedAccounts'].append(account_to_add)

        # Now, you need to update the configuration in AppConfig. 
        # Note: AWS AppConfig does not have direct 'update' methods. 
        # Typically, you'd create a new version of the configuration with the updated data.

        # For the sake of simplicity in this example, let's assume you have a method 
        # `update_appconfig_configuration()` to handle this.
        update_appconfig_configuration(config_data)

    return {
        'statusCode': 200,
        'body': json.dumps('Account added successfully!')
    }

def update_appconfig_configuration(data):
    # Logic to create a new version of the configuration in AWS AppConfig with the updated data.
    # This might involve first updating the configuration file in an S3 bucket or another source 
    # and then pointing the AppConfig configuration profile to this new version.
    pass
```

### 3. Using AppConfig in your Java code to check allowlist:

3.1. **AWS SDK Dependency**:
Make sure you have the AWS SDK for Java added to your project.

3.2. **Fetching Configuration**:

```java
import com.amazonaws.services.appconfig.AmazonAppConfig;
import com.amazonaws.services.appconfig.AmazonAppConfigClientBuilder;
import org.json.JSONObject;

public class AllowListChecker {
    private final AmazonAppConfig client;

    public AllowListChecker() {
        this.client = AmazonAppConfigClientBuilder.standard().build();
    }

    public boolean isAccountAllowed(String accountId) {
        // Fetch configuration from AppConfig
        String content = client.getConfiguration(new GetConfigurationRequest()
            .withApplication("YourAppConfigApplicationName")
            .withEnvironment("YourEnvironment")
            .withConfiguration("YourConfigurationProfile")
            .withClientId("SomeClientIdentifier"))
            .getContent();

        JSONObject config = new JSONObject(content);
        return config.getJSONArray("allowedAccounts").toList().contains(accountId);
    }
}
```

Now, in your main application logic, whenever you need to check if an account has access, you'd use:

```java
AllowListChecker checker = new AllowListChecker();
if (checker.isAccountAllowed("accountToCheck")) {
    // Grant access
} else {
    // Deny access
}
```

### 4. Secure your configuration:

Ensure that you have proper IAM policies in place. Only the Lambda function should be able to update the AppConfig configuration, and your Java application should only have read access to it.

### 5. Maintenance:

Periodically review the allowlist to ensure that only the relevant accounts are on the list. Also, monitor the Lambda function's execution and errors to catch any issues with the updating process.

This approach provides a dynamic way of controlling access via a centralized configuration, making it easier to manage and monitor.

**Comment by ashstarry** [comment]
Sure! Here's the translation and expansion of your text:

---

**Utilizing AWS AppConfig for AllowList Storage**

Considering that SDC was deprecated by AWS AppConfig in 2018, we will opt to use AWS AppConfig to store the AllowList. The scope of storage, whether it is region-based or global, remains to be determined (TBD).

Upon obtaining the desired accounts using one of the previously discussed methods, we will utilize AWS Lambda to push these accounts to a Configuration Profile in AWS AppConfig. Consequently, this list can be accessed and invoked in the `CreateMeshActivity` within our Frontend Service.

The general steps for this process are as follows:

1. **Obtaining Account Information**: Using one of the proposed methods, gather the AWS Account IDs that need to be on the AllowList.

2. **Lambda Function Deployment**:
    - Develop an AWS Lambda function that accepts the AWS Account IDs as input.
    - The function then interacts with AWS AppConfig to update the Configuration Profile with the new Account IDs.

3. **Integration with AWS AppConfig**:
    - Set up an AWS AppConfig Configuration Profile to hold the AllowList.
    - Ensure that permissions are properly set to allow the Lambda function to write to this AppConfig profile and for the Frontend Service to read from it.

4. **Frontend Service Modification**:
    - In the `CreateMeshActivity` of the Frontend Service, make an API call to AWS AppConfig to fetch the latest version of the AllowList.
    - Implement logic to check if the Account attempting to use the `createMesh` API is in the AllowList. If it's not on the list, deny the operation.

5. **Monitoring and Logging**: Implement monitoring and logging mechanisms to track successful and failed attempts at using the `createMesh` API. This will help in diagnosing any potential issues with the AllowList mechanism and ensuring that only authorized accounts can access the service.

6. **Notification Mechanisms** (Optional): Implement a notification mechanism to inform the operations or development team whenever an update is made to the AllowList via the Lambda function.

---

Note: Make sure that all the components, especially the Lambda function, have appropriate IAM roles and permissions to interact with AWS AppConfig and any other required services. Proper testing should be done in a staging environment before rolling out changes to the production system.

**Comment by ashstarry** [comment]
When AWS Config runs its evaluations, it makes API calls to gather current configuration details. These calls might look like attempts to interact with the resource, but they're simply AWS Config retrieving the current state. AWS Config then compares this current state against the desired configurations to determine if the resource is compliant or non-compliant.

**Comment by ashstarry** [comment]
AWS Config will not call the CreateMesh API. AWS Config's role is to monitor and record configurations of AWS resources; it doesn't create, modify, or delete these resources.

When AWS Config evaluates AWS App Mesh configurations, it makes API calls to gather current configuration details of those resources. These calls are typically Describe* or List* operations, like DescribeMesh or ListMeshes, to retrieve the current state of the resource.

**Comment by ashstarry** [comment]
Summary

What exactly is it that we want to discuss? [2:50](https://otter.ai/u/hIzTjWHVcKU5ZzGX9gScRL7MtwA?tab=summary&t=170s)

The product team still needs to meet with the product team to align on customer experience, requirements, and all the things that will inform the engineering decision.

Quick overview of the website.

What is service trust and service access? [4:55](https://otter.ai/u/hIzTjWHVcKU5ZzGX9gScRL7MtwA?tab=summary&t=296s)

Service trust is a permission that customers in their organizations grant to other AWS services, such as ram, config or artifact customers.

Service trust and service access interchangeably.

How to shut off access to atm except to existing customers, especially for larger customers who might want to create a new AWS account.

How an AWS account organization works.

How do you define who is a customer? [9:20](https://otter.ai/u/hIzTjWHVcKU5ZzGX9gScRL7MtwA?tab=summary&t=560s)

They want to block off great mesh only to existing customers. They don't want to be difficult for existing customers, but they do want to make it impossible for customers to access it.

They have had similar services come to them in the past where they integrate with them.

There is no way to implicitly block customer from using the service, but it will be more painful to do it that way.

The initial proposal was to take accounts and then find out all the other accounts that are in the same organization.

What is the hard cap on account capacity? [13:41](https://otter.ai/u/hIzTjWHVcKU5ZzGX9gScRL7MtwA?tab=summary&t=821s)

The public limit is 5000 accounts in the organization, but they are working on scaling that. The public official limit at this time is 5000.

The timeline for onboarding to several organizations.

How long does it take to implement trust integration? [16:05](https://otter.ai/u/hIzTjWHVcKU5ZzGX9gScRL7MtwA?tab=summary&t=965s)

Timeline for a standard service trust integration is three to six months, but depends on the service and what the service team will need to implement.

Most notifications are delivered within one or two minutes.

The speaker has all the questions and all of the docs in case anything comes up. The speaker will be attending the product office hours.

The normal process of production approval.

**Comment by ashstarry** [comment]
Speaker 1  2:50  
From that homepage, but also be sure to link to anything specific that we discussed. So what did you want to I guess I should just start by asking what exactly is it that you wanted to discuss? Because I know we talked a little bit about your product seen your product team still needing to meet with our product team to align on customer experience and requirements and all those things that will then inform the engineering decision or the engineering design and decision and all that.

Speaker 2  3:17  
Sure. Yeah. Thank you for that any hurry. Yeah, I tried this website. So quick overview for our class that we want to make an internal design. So we will get a list of the organization ID from our customer. And these customers, they own the mesh inside of this organization. And what we want is that wherever this organization's gets, new accounts added or created, will want to get detected, for example, from our lambda or app config. And I do notice that from your websites, there is a service called a notification. So we can for example, create an SQS queue and get to notified from the SNS topic you created for us. So we have some quick questions here like first, since it's sad we what we need to our it's better for us to integrate the app and bash our surveys into AWS organization. But actually, even though it is a great integration, we do not have direct feature or director requirement to combine with aqmesh for onboarding, we only want to this application. So in that case, is it possible for us kind of skip the integration process or fast and process

Speaker 1  4:51  
so now for as the I'm not gonna say one way or the other, you can skip the whole integration process. What I am gonna say is this. Most of the services that we integrate, and all of the integrated services that you would see in our public documentation we use something called Service Trust. So Service Trust is a permission that customers in their organizations grant two other AWS services things like RAM or config or artifact. Customers they recall we use the term Service Trust and service access interchangeably and when a customer enables service access for a service, let's say Ram. Ram then begins receiving these notifications. I'm also going to like it sounds like you found the notifications documentation already but for good measure, I'll make sure that okay, I didn't actually have it open for good measure. I'll make sure that you have it here. That will allow your service allows you know RAM or adventure wherever service to start receiving notifications about organizations that have service trust me

Unknown Speaker  6:07  
the service it's, it's

Speaker 1  6:23  
giving you there already, again, something that I'm already hearing myself and realizing it's more of a product discussion, and more of something that you know, our product team will be able to explain your product even align more on what the expectations would be from like integrating with organizations perspective. But what I'll say again, what I'll say is this we can accomplish what you're trying to accomplish. It's just I can't say one way or the other. If we could give you an end around to the typical service trust integration from what you've explained. It sounds very much like something that would fit within normal service trust integration, but again, that's a decision that we can make at the profit level.

Speaker 3  7:06  
I airy so I'm I'm then also from that team. So what we're trying to do is we want to shut off access of ATM except to existing customers. And one of the questions that had come up was how do we do it in a way especially for larger customers who might want to create a new AWS account and still be able to access it without like, having to go through like cutting a support ticket or something like that? So we were wondering, is there a way using like an OBS organization kind of just behind the scenes not really involving the customer, like having them like explicitly offering them? Like opting in, just be able to like, kind of like, auto detect, like hey, you know, we've seen accounts are this organization that I've used that match just grants, other accounts in our organization, as well. So if I'm understanding

Speaker 1  8:08  
correctly, you want to you don't you actually almost myself, don't want to integrate data resources. This follows kind of something you want. My quite the exact opposite, but what you want to do is automatically, like detect accounts that are using aqmesh and are in organizations and by that you want to allow the other accounts in in the same organization, as the accountants using App punish us out of match. But other organizations, all of those accounts you want to disallow them from using

Speaker 3  8:46  
aqmesh. Yeah, so we want to restrict that match basically, the existing customers

Speaker 1  8:53  
how does the non and how does it How does it How does it I have a lot of questions that I might just have to kick the can down the road to like this is going to end up being a product discussion, but I'm just gonna know how exactly that I don't know. How exactly does an AWS account organization or not go from not an average customer to an average customer? I think I might need to understand that. I would say it would go

Speaker 3  9:20  
who would become my customer by going in and they log into their account. And they like call that match create mesh endpoint. And that's kind of how we would say that they are now a customer. But we want to block off great mesh only to existing customers. It's like we could like we could just pull out a list of account IDs, but the problem with that is like if a new account IDs for like a big customer like Starbucks transmigrating match we want to be able to select them credo without too much friction on their side or our side.

Speaker 1  9:55  
It sounds like you do want to just want to understand that.

Speaker 3  10:00  
It's complicated we want we definitely don't want it to be as easy, especially for customers. Well, we don't want to be difficult for customers that are existing, but like we do want to be impossible for customers. I've never had that much at all, because account IDs don't really represent like customers. You Yeah, even get an AWS where we have like tons and tons of account IDs just for Yeah.

Speaker 1  10:26  
So it's funny because we've had, we've had similar senator, we've had similar services come to us in the past where they integrate with us but they are blocked off to like a specific subset of currently they only want to allow like very specific customers what what I'll say is this is like your right on when you say like an account and use an account ID we don't know like how that maps in terms of like a higher level like business customer perspective. I don't want to go down into this areas organizations doesn't have a clue like are we have no we have no we don't have any like implicit visibility into like which account I use belong to again, like which business level customer that's something if we wanted to we can figure it out, but the service in and of itself does not store that data nor does it like really have easy like I should say programmatic access to it. So that's I guess, something we would not be inherently be able to offer. And there would be like I said we had we've supported services different services like this in different ways. Before, there wouldn't be like an implicit there wouldn't be a way where we can like implicitly and functionally block customer from using our service, but I'll say is this it's not it's going to be more painful for you and it's going to be more painful for us to do it that way then it probably wouldn't be for your service. To find a way to implement that. mapping of again accounts to it, isn't it your business level customer I don't really know if that's right or not. But that's kind of how I'm thinking about it within your own service because the services that we've supported in the past that have wanted to be restricted to like a very limited set of customers have been like like a service that was like only being stood up for like the Singaporean government or something like that. So we're talking like very, very, very, very, very, very people. And again, I do want to route back to you this is all like audit level discussion like I actually have a product manager I'm gonna show you I'm not but just speaking about the functional perspective. eyeballing what the work would look like based on what we've talked about. Yeah, that's just that's just my first impression of it all.

Speaker 3  12:41  
Yeah, I totally understand that, like a didn't seem like the most perfect set, but at the same time, like we did want to try to find something that was more than just here's a list of a lot of IDs, but like yeah, this is not like an approach that makes us that's fine.

Speaker 1  13:06  
It's just gonna be hard because of the I mean, I sent you a little bit about Service Trust. Isn't there many reasons it's set up that way because it's set up that way that kind of like I guess the initial the initial proposal you brought, just really you want to take accounts and then find out all the other accounts that are in the same organization as that account are allowed to but any other accounts are not allowed to that just that just kind of like you're saying it doesn't really align with the way how we define and how we actually implemented our actual like service permission boundary, which is called service account customers.

Speaker 3  13:41  
Gotcha. Okay. That's good to know. Yeah, that's, that's good. All the questions I have Brian go. Any thoughts? Sounds like a pretty good answer for what we're looking for.

Speaker 1  14:06  
Awesome. I'm glad to hear it and I do encourage like, definitely follow up, follow up like, you know, follow through with our product. I think you're on our next webinar scheduled for our next product office hours. So all root then don't take anything I just said, if you know absolutely too hard, they will know a lot more about what we definitely can but we definitely can't support you know, will we be willing just work in different situations they'll be able to understand again, the product and business use cases a lot better than I can again because I'm just a product manager. But they go for joining me it's very interesting to learn a little bit about your service and be able to provide a little bit of it, a little bit of clarity about what our services and could do and maybe unfortunately, unfortunately.

Speaker 2  14:47  
Oh yes, I'm registered for the production office hour and what click traffic referring to that I noticed from the website is say well, organization can hold a solid account, right? So the limit won't be huger blocker here and for example, for one onboarding to several different organizations.

Speaker 1  15:13  
Yeah, how does the accountant and organization thing this is an MSA someone.

Speaker 2  15:17  
So this is what highest find from your official website is saying that that's when organizations can afford five solid accounts.

Speaker 1  15:24  
Yeah. Yeah, that's correct. The public limit is 5000. hard cap. We're working on scaling that I believe there have been exceptions that we've made on very, very strict case by case basis. But yeah, the public official limit at this time was 5000. accounts in the organization.

Speaker 2  15:45  
Sure, and I'm asking you that maybe too early to answer but normally, could we have kind of our rough evaluation for a timeline? Because I saw from the website their timeline graph is sad Ross may need half a year for the whole process.

Speaker 1  16:05  
Or like a standard service trust integration. Yeah, it's broadly speaking three to six months. There are many things that it depends on and honestly it all ends up depending on the service and what the service team will actually need to implement. And because your service will probably, again, from the sound, it might not be pursuing a standard service trust integration. Like that's, again, just the timeline I was able to live with a great sense of might not end up applying.

Unknown Speaker  16:33  
Yeah, and so

Speaker 1  16:37  
but in a general sense, yeah, we ended up looking at like a several month timeline anywhere from three to six months. Again, the broadly speaking depends on different factors. A lot of that time mainly being like the service team, building how and testing and you know, doing pre launch stuff.

Speaker 2  17:02  
Sure, so maybe more related to the implementation. So I have one small staff that I notice for the differences especially high for the notification. And I noticed there is one called Add Account to organization and it's a it is async. So which means it will take roughly 15 minutes, but ideally it within one minute or two minutes for us to receive that affiliation, so that normally is the case, right and the hub other reality.

Speaker 1  17:36  
Yeah, so the documentation says we keep an SLA of 15 minutes but most notifications that are delivered within one or two minutes.

Speaker 2  17:46  
Okay, good to know I have all my questions that

Speaker 1  17:52  
I, you have all of our docs in case anything comes up, you will be attending our product office hours so I get a lot a lot of this stuff. I tried to say take a stab at hopefully this more concrete discussion on that on Thursday or whatever. I don't know I cannot really keep track of trying to keep track of the days anymore. Cool. Any other questions? Feel free to speak now. Otherwise, you can follow up. I think you do have a ticket with us. So any engineering based questions you would ask them there but I'll differ a lot for product managers for your upcoming partnerships our session.

Speaker 3  18:51  
All right. I was just saying I'm hearing a lot of silence. I guess we're gonna end things here.

Speaker 1  19:01  
Yes, thank you all for joining and we'll be sure if I'll follow up with our product managers and let them know that you'll be joining on you'll be sure to join in your session, maybe a little about some little bit about some of the things that you brought up say this department.

Speaker 2  19:20  
Sure, yeah. I cannot take this for the production when I notice for the normal process of production approval first then we communicate with engineer before the detailed implementation.

Unknown Speaker  19:39  
Thanks for your help. Yeah.

Speaker 2  19:51  
Yeah, exactly. If there's no other questions, I would have sent you offline. Yes. Thank you. For help.

Unknown Speaker  19:56  
Sounds good country. Thank you. Thanks.

Transcribed by https://otter.ai

**Comment by ashstarry** [comment]
We can reduce the default mesh quota for accounts from 15 to 0, but we will not alter the mesh quota for accounts on the allowlist, thereby indirectly achieving the goal of preventing new customers from creating meshes. To ensure a smoother transition and reduce operational workload, in the initial phase after announcing the discontinuation of new customer onboarding, if customers genuinely need to create a mesh and request a quota increase, we can set it for auto-approval, but we will log their account details.

**Comment by ashstarry** [comment]
"This is not a typical use case for service quotas, and users might misunderstand the operational process. As there's no precedent for such a setup, we may need to communicate with the Minerva team."

**Comment by ashstarry** [comment]
We will choose AWS AppConfig (its status in the China region is GA, but we need to reconfirm its availability) or DynamoDB as the storage solution for this dynamically expanded allowlist. It will be integrated with the Frontend in the form of a feature flag.
In general, if we choose a solution that does not require interfacing with external teams, we can complete the external closure of the AppMesh entry within one month.

**Comment by ashstarry** [comment]
```
import os

# Directory containing your files
directory = '/path/to/your/folder'

# Initialize an empty set to store unique AWS account IDs
account_ids = set()

# Iterate over each file in the directory
for filename in os.listdir(directory):
    # Build full path to file
    filepath = os.path.join(directory, filename)
    
    # Ensure it's a file and not a directory or symlink
    if os.path.isfile(filepath):
        with open(filepath, 'r') as file:
            # Skip the first line (title)
            next(file)
            
            # Add each account ID to the set
            for line in file:
                account_ids.add(line.strip())

# Convert the set to a list
unique_account_ids = list(account_ids)

# Optional: Save the consolidated list to a new file
with open('consolidated_account_ids.txt', 'w') as outfile:
    for account_id in unique_account_ids:
        outfile.write(account_id + '\n')

# Or simply print the unique account IDs
# for account_id in unique_account_ids:
#     print(account_id)

```
